### ISSUE TITLE
Problem with a custom object and 'instanceof'.

### ISSUE THREAD

**Issue by adudov** [Microsoft/ChakraCore#5127]
We have the following setup: (the objects are created but 'instanceof' does not work)

```
JsValueRef CHAKRA_CALLBACK CommentConstructor(
    JsValueRef callee,
    bool isConstructCall,
    JsValueRef* arguments,
    unsigned short argumentCount,
    void* callbackState)
{
    JsValueRef jsObject;
    // Creating the jsObject object with the prototype for the Comment elements. (via JsCreateExternalObjectWithPrototype)
    return jsObject;
}


JsValueRef callback = JS_INVALID_REFERENCE;
JsCreateNamedFunction(functionName, &CommentConstructor, nullptr, &callback);

JsValueRef jsPropDesc = JS_INVALID_REFERENCE;
JsCreateObject(&jsPropDesc);
SetChakraCoreProperty(jsPropDesc, "enumerable", MakeChakraCoreBooleanValue(true));
SetChakraCoreProperty(jsPropDesc, "value", callback);

JsValueRef globalObj = JS_INVALID_REFERENCE;
JsGetGlobalObject(&globalObj);
DefineChakraCoreProperty(globalObj, "Comment", jsPropDesc);
```
(where `SetChakraCoreProperty` just sets the property, `DefineChakraCoreProperty` defines the property with the descriptor and `MakeChakraCoreBooleanValue` creates a js bool variable)

In JavaScript we have the following:
```
var comm = new Comment();
var isInstanceOf = comm instanceof Comment;
```
Which results to 'false'.
What is the correct way of creating and setting an object's constructor (it's prototype)?

**Comment by jackhorton** [comment]
@MSLaguana and I discussed this and realized that it doesn't seem like this is possible today. We should be able to add some information to JsCreateEnhancedFunction callbacks to get the info needed for this, though.

**Comment by MSLaguana** [comment]
More specifically, it's not possible in an easy way.
What you need to do is set `Comment.prototype` as the prototype of the `jsObject` in your constructor, but the native constructor isn't given the js function object so there is a bit of a chicken-and-egg problem. You could pre-initialize your `Comment.prototype` object and pass a reference to that in the `callbackState` argument, and then after you `JsCreateNamedFunction` you could install it as the `Comment.prototype`, and I think that would work.

It would be much nicer if we just passed a reference to the javascript function in as part of the `JsNativeFunctionInfo` given to native callbacks.

**Comment by rhuanjl** [comment]
My friend @fatcerberus does this in his miniSphere game engine https://github.com/fatcerberus/minisphere
See this function:
https://github.com/fatcerberus/minisphere/blob/master/src/shared/api.c#L333-L392

It's a bit hard to follow directly as he wraps all of his Jsrt calls in an abstraction jsal layer designed to mimic duktape's api (except as jsal_methodName instead of duk_methodName) (as the engine was originally designed to run with duktape and then ported to use CC instead)

Roughly though he does something like this:
1. make a prototype object
2. make the constructor function
3. Set prototype.constructor to the constructor function (using JsDefineProperty)
4. Set constructor.prototype to the prototype (using JsDefineProperty)

**Comment by adudov** [comment]
Thank you all for the fast response.
I tried the @rhuanjl's suggestion and it worked.

**Comment by MSLaguana** [comment]
I just realized that I misspoke before: The `callee` argument to the native function is the javascript function, so you could access `callee.prototype` and set that as the prototype of the constructed object, then you don't need to set up the prototype object up front if you don't want to.

**Comment by fatcerberus** [comment]
In my experience it’s better to use JsCreateExternalObjectWithPrototype from the get-go than to set the prototype afterwards; the latter is quite expensive.  I found this out when optimizing my own codebase, in native constructor-heavy code I was spending a significant amount of time in JsSetPrototype, switching to CreateObjectWithPrototype in these cases granted a significant speedup.
